generator/src/codegen/cpp.rs

```diff
@@ -0,0 +1,385 @@
+        self.indent();
+        for elem in r.contents() {
+            match elem {
+                RestrictionContent::Enumeration(e) => {
+                    if let Some(doc) = e.doc().as_ref() {
+                        cg!(self, "/* {} */", doc);
+                    }
+                    cg!(self, "{} = {},", e.value().to_shouty_snake_case(), e.id());
+                }
+            }
+        }
+        self.dedent();
+        cg!(self, "}};");
+        Ok(())
+    }
+
+    fn write_simple_type_header(&mut self, s: &SimpleType) -> Result<()> {
+        if let Some(doc) = s.doc() {
+            cg!(self, "/* {} */", doc);
+        }
+        match s.content() {
+            SimpleTypeContent::Restriction(r) => self.write_restriction(s.name(), r)?,
+        }
+        Ok(())
+    }
+
+    fn write_simple_type_source(&mut self, _: &SimpleType) -> Result<()> {
+        Ok(())
+    }
+
+    fn write_simple_type(&mut self, s: &SimpleType) -> Result<()> {
+        self.select(Select::H);
+        self.write_simple_type_header(s)?;
+        self.select(Select::CPP);
+        self.write_simple_type_source(s)
+    }
+
+    fn write_complex_type_header(&mut self, c: &ComplexType) -> Result<()> {
+        if let Some(doc) = c.doc() {
+            cg!(self, "/* {} */", doc);
+        }
+        match c.content() {
+            ComplexTypeContent::Choice(choice) => self.write_choice_header(c.name(), choice)?,
+            ComplexTypeContent::Seq(s) => self.write_sequence_header(c.name(), s)?,
+            ComplexTypeContent::Empty => {
+                self.select(Select::H);
+                cg!(self, "struct {} {{}};", c.name().to_camel_case());
+            }
+        }
+        Ok(())
+    }
+
+    fn write_complex_type_source(&mut self, c: &ComplexType) -> Result<()> {
+        match c.content() {
+            ComplexTypeContent::Choice(choice) => self.write_choice_source(c.name(), choice)?,
+            ComplexTypeContent::Seq(s) => self.write_sequence_source(c.name(), s)?,
+            ComplexTypeContent::Empty => {}
+        }
+        Ok(())
+    }
+
+    fn write_complex_type(&mut self, c: &ComplexType) -> Result<()> {
+        self.select(Select::H);
+        self.write_complex_type_header(c)?;
+        self.select(Select::CPP);
+        self.write_complex_type_source(c)
+    }
+}
+
+impl<'a> Codegen for Generator<'a> {
+    fn generate(&mut self, packet: &Packet) -> Result<Vec<String>> {
+        let mut vec = Vec::new();
+        vec.push(self.generate_header()?);
+        vec.push(self.generate_source()?);
+
+        self.select(Select::H);
+        cg!(self, "/* Generated with IDL v{} */\n", self.version);
+        cg!(self,"#include <vector>");
+        cg!(self, "#include <string>");
+        cg!(self, "#include <array>");
+        cg!(self, "#include <variant>");
+        cg!(self, "#include <string_view>");
+        cg!(self, "#include <optional>");
+        cg!(self);
+        cg!(self, "namespace Packet {{");
```

Comment:
What double `{{`?

Reply:
it's calling `format!()` behind the scenes, so we need the double `{{` otherwise the compiler will complain